From: Matthew Hagemann <matt.hagemann@canonical.com>
Date: Thu, 13 Feb 2025 17:08:34 +0200
Subject: Dynamic lang pack installation support

Adding a polkit rule to allow the gnome initial setup user to run pkexec if the
program it is running is apt. This is to allow for Gnome-Initial-Setup install
language packs during setup via apt update && apt install.

Adding a function apt_install that takes in a list of packages and attempts to
install them. This is inteded to be used as part of the new flow for
dynamically installing language packs for the user during setup.

Added parse_language_list function to parse the static list of languages in
/usr/share/language-selector/data/languagelist which is now used to populate
the language chooser. Gnome lang packs are used as a fallback.

Added the ability for GIS to determine what outstanding language packs need to
be installed via check-language-support -l and to use that output to install
outstanding lang packs via a g_task.

Additional language packs are installed at the end of the flow when the user
lands on the summary page. A spinner and prompt is shown that indicates lang
packs are being downloaded. Once complete, the user can start using Ubuntu.
They can optionally cancel the download and proceed to Ubuntu. No download is
attempted if there is no active network connection.

Additional PR style updates

PR code style cleanup
---
 data/20-gnome-initial-setup.rules.in               |  10 +
 data/gnome-initial-setup-wrapper                   |   1 +
 gnome-initial-setup/gis-pkexec.c                   |  65 +++++-
 gnome-initial-setup/gis-pkexec.h                   |   6 +
 .../pages/language/cc-language-chooser.c           |  47 ++++-
 .../pages/software/gis-software-page.c             |   2 +-
 .../pages/summary/gis-summary-page.c               | 232 +++++++++++++++++++--
 .../pages/summary/gis-summary-page.ui              |  53 ++++-
 8 files changed, 386 insertions(+), 30 deletions(-)

diff --git a/data/20-gnome-initial-setup.rules.in b/data/20-gnome-initial-setup.rules.in
index 02fd21d..d769020 100644
--- a/data/20-gnome-initial-setup.rules.in
+++ b/data/20-gnome-initial-setup.rules.in
@@ -5,6 +5,16 @@
 // Allow the gnome-initial-setup user to do certain actions without
 // being interrupted by password dialogs
 
+polkit.addRule(function(action, subject) {
+    if (subject.user !== 'gnome-initial-setup')
+        return undefined;
+
+    if (action.id === 'org.freedesktop.policykit.exec' &&
+        action.lookup('program') === '/usr/bin/apt-get') {
+        return 'yes';
+    }
+});
+
 polkit.addRule(function(action, subject) {
     if (subject.user !== 'gnome-initial-setup')
         return undefined;
diff --git a/data/gnome-initial-setup-wrapper b/data/gnome-initial-setup-wrapper
index 0e276f5..5be7c08 100755
--- a/data/gnome-initial-setup-wrapper
+++ b/data/gnome-initial-setup-wrapper
@@ -1,6 +1,7 @@
 #!/bin/bash
 
 CONVD_WRAPPER=/usr/libexec/launch-desktop-provision-init
+ulimit -n 4096
 
 if [ -f "${CONVD_WRAPPER}" ]; then
 # call convd wrapper passing as first parameter the full path to
diff --git a/gnome-initial-setup/gis-pkexec.c b/gnome-initial-setup/gis-pkexec.c
index 9e1bfa9..6654b006 100644
--- a/gnome-initial-setup/gis-pkexec.c
+++ b/gnome-initial-setup/gis-pkexec.c
@@ -29,6 +29,7 @@ gboolean
 gis_pkexec (const gchar  *command,
             const gchar  *arg1,
             const gchar  *user,
+            GCancellable *cancellable,
             GError      **error)
 {
   g_autoptr(GSubprocessLauncher) launcher = NULL;
@@ -43,6 +44,8 @@ gis_pkexec (const gchar  *command,
    * so remove it from the environment.
    */
   g_subprocess_launcher_unsetenv (launcher, "SHELL");
+  g_subprocess_launcher_setenv (launcher, "DEBCONF_NONINTERACTIVE_SEEN", "true", TRUE);
+  g_subprocess_launcher_setenv (launcher, "DEBIAN_FRONTEND", "noninteractive", TRUE);
   process = g_subprocess_launcher_spawnv (launcher, argv, error);
 
   if (!process) {
@@ -50,10 +53,70 @@ gis_pkexec (const gchar  *command,
     return FALSE;
   }
 
-  if (!g_subprocess_wait_check (process, NULL, error)) {
+  if (!g_subprocess_wait_check (process, cancellable, error)) {
     g_prefix_error (error, "%s failed: ", command);
     return FALSE;
   }
 
   return TRUE;
 }
+
+gboolean
+apt_get_install (GPtrArray     *packages,
+                 GCancellable  *cancellable,
+                 GError       **error) {
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  static gboolean already_updated = FALSE;
+
+  g_return_val_if_fail (packages != NULL, FALSE);
+  g_return_val_if_fail (packages->len != 0, FALSE);
+
+  if (!g_atomic_int_get (&already_updated)) {
+    /* Avoid running apt-update again if it has been already completed in a previous iteration */
+    if (!gis_pkexec ("/usr/bin/apt-get", "update", NULL, cancellable, error))
+      return FALSE;
+
+    g_atomic_int_set (&already_updated, TRUE);
+  }
+
+  /* Create space for 8 leading arguments of the install command, the number of
+   * packages and a NULL terminator.
+   */
+  int total = 4 + packages->len + 1;
+  g_autoptr (GPtrArray) argv = g_ptr_array_new_null_terminated (total, g_free, TRUE);
+
+  g_ptr_array_add (argv, g_strdup ("/usr/bin/pkexec"));
+  g_ptr_array_add (argv, g_strdup ("/usr/bin/apt-get"));
+  g_ptr_array_add (argv, g_strdup ("install"));
+  g_ptr_array_add (argv, g_strdup ("-y"));
+  while (packages->len) {
+    g_ptr_array_add (argv, g_ptr_array_steal_index_fast (packages, 0));
+  }
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_unsetenv (launcher, "SHELL");
+
+  g_subprocess_launcher_setenv (launcher, "DEBCONF_NONINTERACTIVE_SEEN", "true", TRUE);
+  g_subprocess_launcher_setenv (launcher, "DEBIAN_FRONTEND", "noninteractive", TRUE);
+
+  process =
+      g_subprocess_launcher_spawnv (launcher, (const gchar *const *) argv->pdata, error);
+
+  if (!process) {
+    g_prefix_error_literal (error, "Failed to create apt install process: ");
+    return FALSE;
+  }
+
+  if (g_cancellable_set_error_if_cancelled (cancellable, error)) {
+    g_prefix_error_literal (error, "User canceled subprocess: ");
+    return FALSE;
+  }
+
+  if (!g_subprocess_wait_check (process, cancellable, error)) {
+    g_prefix_error_literal (error, "apt-get install failed: ");
+    return FALSE;
+  }
+
+  return TRUE;
+}
diff --git a/gnome-initial-setup/gis-pkexec.h b/gnome-initial-setup/gis-pkexec.h
index ff5a88c..000bea9 100644
--- a/gnome-initial-setup/gis-pkexec.h
+++ b/gnome-initial-setup/gis-pkexec.h
@@ -34,8 +34,14 @@ gboolean
 gis_pkexec (const gchar  *command,
             const gchar  *arg1,
             const gchar  *user,
+            GCancellable *cancellable,
             GError      **error);
 
+gboolean
+apt_get_install (GPtrArray     *packages,
+                 GCancellable  *cancellable,
+                 GError       **error);
+
 G_END_DECLS
 
 #endif /* __GIS_PKEXEC_H__ */
diff --git a/gnome-initial-setup/pages/language/cc-language-chooser.c b/gnome-initial-setup/pages/language/cc-language-chooser.c
index 61756be..d68d33b 100644
--- a/gnome-initial-setup/pages/language/cc-language-chooser.c
+++ b/gnome-initial-setup/pages/language/cc-language-chooser.c
@@ -21,6 +21,7 @@
  */
 
 #include "config.h"
+#include "gis-pkexec.h"
 #include "cc-language-chooser.h"
 
 #include <locale.h>
@@ -30,6 +31,8 @@
 #include <gtk/gtk.h>
 
 #define GNOME_DESKTOP_USE_UNSTABLE_API
+#define LANGUAGE_LIST_PATH "/usr/share/language-selector/data/languagelist"
+
 #include <libgnome-desktop/gnome-languages.h>
 
 #include "cc-common-language.h"
@@ -207,7 +210,6 @@ sync_all_checkmarks (CcLanguageChooser *chooser)
 
                 should_be_visible = g_str_equal (widget->locale_id, priv->language);
                 gtk_widget_set_opacity (widget->checkmark, should_be_visible ? 1.0 : 0.0);
-
                 row = gtk_widget_get_next_sibling (row);
         }
 }
@@ -290,13 +292,54 @@ add_languages (CcLanguageChooser  *chooser,
         gtk_list_box_set_placeholder (GTK_LIST_BOX (priv->language_list), priv->no_results);
 }
 
+static GStrv
+parse_language_list (void)
+{
+    g_autofree gchar *contents = NULL;
+    g_autoptr (GError) error = NULL;
+
+    if (!g_file_get_contents (LANGUAGE_LIST_PATH, &contents, NULL, &error)) {
+        g_critical ("Error reading file: %s", error->message);
+        return NULL;
+    }
+
+    g_auto (GStrv) lines = g_strsplit (contents, "\n", -1);
+    g_autoptr (GStrvBuilder) builder = g_strv_builder_new ();
+
+    for (gchar **line = lines; *line != NULL; line++) {
+        const gchar *trimmed = g_strstrip (*line);
+
+        /* Skip empty lines and commented out lines */
+        if (trimmed[0] == '\0' || trimmed[0] == '#') {
+            continue;
+        }
+
+        g_auto (GStrv) tokens = g_strsplit (trimmed, ";", -1);
+        gint token_count = g_strv_length (tokens);
+
+        /* Language code we use is the 4th entry in a ; delimited list */
+        if (token_count >= 4)
+            g_strv_builder_add (builder, g_strdup (tokens[3]));
+    }
+
+    return g_strv_builder_end (builder);
+}
+
 static void
 add_all_languages (CcLanguageChooser *chooser)
 {
         g_auto(GStrv) locale_ids = NULL;
         g_autoptr(GHashTable) initial = NULL;
 
-        locale_ids = gnome_get_all_locales ();
+        /* Make use of the static language list in
+         * /usr/share/language-selector/data/languagelist. If not present, fall
+         * back to using gnome_get_all_locales().
+         */
+        if (!(locale_ids = parse_language_list ())) {
+                g_warn_if_reached ();
+                locale_ids = gnome_get_all_locales ();
+        }
+
         initial = cc_common_language_get_initial_languages ();
         add_languages (chooser, locale_ids, initial);
 }
diff --git a/gnome-initial-setup/pages/software/gis-software-page.c b/gnome-initial-setup/pages/software/gis-software-page.c
index 0e2278f..6892365 100644
--- a/gnome-initial-setup/pages/software/gis-software-page.c
+++ b/gnome-initial-setup/pages/software/gis-software-page.c
@@ -96,7 +96,7 @@ gis_software_page_apply (GisPage      *gis_page,
       program = find_fedora_third_party ();
       if (program)
         {
-          gis_pkexec (program, "enable", "root", &error);
+          gis_pkexec (program, "enable", "root", cancellable, &error);
           if (error && !g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
             g_warning ("%s failed: %s", program, error->message);
         }
diff --git a/gnome-initial-setup/pages/summary/gis-summary-page.c b/gnome-initial-setup/pages/summary/gis-summary-page.c
index ba9e1fb..64a2745 100644
--- a/gnome-initial-setup/pages/summary/gis-summary-page.c
+++ b/gnome-initial-setup/pages/summary/gis-summary-page.c
@@ -21,6 +21,9 @@
 
 /* Summary page {{{1 */
 
+#include "adwaita.h"
+#include "glib.h"
+#include <locale.h>
 #define PAGE_ID "summary"
 
 #include "config.h"
@@ -39,10 +42,16 @@
 
 struct _GisSummaryPagePrivate {
   GtkWidget *start_button;
+  GtkWidget *spinner;
+  GtkWidget *cancel_button;
+  GtkWidget *status_stack;
+
   AdwStatusPage *status_page;
 
   ActUser *user_account;
   const gchar *user_password;
+
+  GCancellable *install_cancellable;
 };
 typedef struct _GisSummaryPagePrivate GisSummaryPagePrivate;
 
@@ -195,32 +204,128 @@ done_cb (GtkButton *button, GisSummaryPage *page)
     }
 }
 
-static void
-gis_summary_page_shown (GisPage *page)
+static GPtrArray *
+get_lang_support_packages_for_locale (const gchar   *locale,
+                                      GCancellable  *cancellable,
+                                      GError       **error)
 {
-  GisSummaryPage *summary = GIS_SUMMARY_PAGE (page);
-  GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (summary);
-  g_autoptr(GError) local_error = NULL;
+        g_autoptr (GSubprocess) subprocess = NULL;
+        g_autofree gchar *stdout_buf = NULL;
+        g_autofree gchar *stderr_buf = NULL;
+
+        subprocess = g_subprocess_new (G_SUBPROCESS_FLAGS_STDOUT_PIPE |
+                                       G_SUBPROCESS_FLAGS_STDERR_PIPE,
+                                       error,
+                                       "/usr/bin/check-language-support",
+                                       "-l",
+                                       locale,
+                                       NULL);
+
+        if (!subprocess)
+                return NULL;
+
+        if (!g_subprocess_communicate_utf8 (subprocess, NULL, cancellable,
+                                            &stdout_buf, &stderr_buf, error)) {
+                g_prefix_error_literal (error, "Failed to communicate with check-language-support: ");
+                return NULL;
+        }
+
+ 
+
+        if (!g_subprocess_wait_check (subprocess, cancellable, error)) {
+                g_prefix_error_literal (error, "check-language-support returned an error: ");
+                return NULL;
+        }
+
+        if (stderr_buf && *stderr_buf != '\0')
+                g_warning ("check-language-support stderr: %s", stderr_buf);
+
+        if (!stdout_buf || *stdout_buf == '\0')
+                return g_ptr_array_new_with_free_func (g_free);
+
+        g_auto (GStrv) packages_split = g_strsplit (g_strstrip (stdout_buf), " ", -1);
+        return g_ptr_array_new_take_null_terminated ((gpointer *) g_steal_pointer (&packages_split), g_free);
+}
 
-  if (!gis_driver_save_data (GIS_PAGE (page)->driver, &local_error))
-    {
-      g_warning ("Error saving data: %s", local_error->message);
+static gboolean
+install_language_support (const gchar   *locale,
+                          GCancellable  *cancellable,
+                          GError       **error)
+{
+        g_autoptr (GPtrArray) packages = NULL;
+
+        g_autofree gchar *locale_for_install = NULL;
+        char *suffix = strchr (locale, '.');
+        if (suffix)
+                locale_for_install = g_strndup (locale, suffix - locale);
+        else
+                locale_for_install = g_strdup (locale);
+
+        packages = get_lang_support_packages_for_locale (locale_for_install, cancellable, error);
+        if (!packages)
+                return FALSE;
+
+        if (packages->len == 0) {
+                g_message ("No packages to install for locale: %s", locale_for_install);
+                return TRUE;
+        }
+
+        if (!apt_get_install (packages, cancellable, error)) {
+                g_prefix_error_literal (error, "Error installing language support: ");
+                return FALSE;
+        }
+
+        return TRUE;
+}
 
-      GtkWindow *parent = GTK_WINDOW (gtk_widget_get_root (GTK_WIDGET (page)));
-      GtkWidget *dialog = adw_message_dialog_new (parent,
-                                                  _("Setup Failed"),
-                                                  local_error->message);
-      adw_message_dialog_add_response (ADW_MESSAGE_DIALOG (dialog), "close", _("Close"));
-      /* FIXME: Provide some more options for debugging or recovery */
+static gboolean
+check_network_connection(void)
+{
+    g_autoptr (GNetworkMonitor) monitor = g_network_monitor_get_default();
+    return g_network_monitor_get_network_available(monitor);
+}
 
-      gtk_window_present (GTK_WINDOW (dialog));
-    }
+static void
+installation_task_async (GTask        *task,
+                         gpointer      source_object,
+                         gpointer      task_data,
+                         GCancellable *cancellable)
+{
+        const gchar *locale = task_data;
+        g_autoptr (GError) error = NULL;
+        
+        /* Skip installation attempt if there is no active connection. */
+        if (!check_network_connection()) {
+                g_task_return_boolean(task, TRUE);
+                return;
+        }
+
+        if (g_cancellable_set_error_if_cancelled (cancellable, &error)) {
+                g_task_return_error (task, g_steal_pointer(&error));
+                return;
+        }
+
+        if (!install_language_support (locale, cancellable, &error)) {
+                g_task_return_error (task, g_steal_pointer (&error));
+                return;
+        }
+
+        g_task_return_boolean(task, TRUE);
+
+ 
+}
 
-  gis_driver_get_user_permissions (GIS_PAGE (page)->driver,
-                                   &priv->user_account,
-                                   &priv->user_password);
+static void
+update_status_stack (GisSummaryPage    *page,
+                     const gchar       *error_message)
+{
+        GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (page);
+        
+        if (error_message) {
+            g_warning("Error: %s", error_message);
+        }
 
-  gtk_widget_grab_focus (priv->start_button);
+        gtk_stack_set_visible_child(GTK_STACK(priv->status_stack), priv->start_button);
 }
 
 static void
@@ -248,6 +353,89 @@ update_distro_name (GisSummaryPage *page)
   g_free (text);
 }
 
+static void
+installation_task_done (GObject      *source_object,
+                        GAsyncResult *res,
+                        gpointer      user_data)
+{
+        g_autoptr (GError) error = NULL;
+        GisSummaryPage *page = user_data;
+        GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (page);
+
+        if (g_task_propagate_boolean (G_TASK (res), &error)) {
+                update_status_stack (page, FALSE, NULL);
+                update_distro_name(page);
+                adw_status_page_set_title(priv->status_page, _("All done!"));
+                return;
+        }
+
+        if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+                update_status_stack (page, FALSE,
+                                     _("Canceled additional downloads"));
+                return;
+        }
+
+        update_status_stack (page, FALSE,
+                             _("Unable to download additional language packs"));
+
+        g_warning ("Installation failed: %s", error->message);
+}
+
+static void
+gis_summary_page_shown (GisPage *page)
+{
+  GisSummaryPage *summary = GIS_SUMMARY_PAGE (page);
+  GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private (summary);
+  g_autoptr(GError) local_error = NULL;
+
+  adw_status_page_set_description(priv->status_page, "");
+
+  adw_status_page_set_title(priv->status_page, _("Nearly done"));
+
+  gtk_stack_set_visible_child_name(GTK_STACK(priv->status_stack), "spinner_box");
+  gtk_spinner_start (GTK_SPINNER (priv->spinner));
+
+  if (!gis_driver_save_data (GIS_PAGE (page)->driver, &local_error))
+    {
+      g_warning ("Error saving data: %s", local_error->message);
+
+      GtkWindow *parent = GTK_WINDOW (gtk_widget_get_root (GTK_WIDGET (page)));
+      GtkWidget *dialog = adw_message_dialog_new (parent,
+                                                  _("Setup Failed"),
+                                                  local_error->message);
+      adw_message_dialog_add_response (ADW_MESSAGE_DIALOG (dialog), "close", _("Close"));
+      /* FIXME: Provide some more options for debugging or recovery */
+
+      gtk_window_present (GTK_WINDOW (dialog));
+    }
+
+  gis_driver_get_user_permissions (GIS_PAGE (page)->driver,
+                                   &priv->user_account,
+                                   &priv->user_password);
+
+  g_cancellable_cancel (priv->install_cancellable);
+  priv->install_cancellable = g_cancellable_new ();
+
+  g_autoptr (GTask) task = g_task_new (page, priv->install_cancellable,
+                                             installation_task_done, page);
+
+  GisDriver *driver = GIS_PAGE(page)->driver;
+  const char *locale = gis_driver_get_user_language(driver);
+  g_task_set_task_data (task, g_strdup (locale), g_free);
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_run_in_thread (task, installation_task_async);
+}
+
+static void
+status_stack_reset_cb (GtkButton *button, GisSummaryPage *page)
+{
+    GisSummaryPagePrivate *priv = gis_summary_page_get_instance_private(page);
+    g_cancellable_cancel(priv->install_cancellable);
+    update_status_stack(page, FALSE, NULL);
+    update_distro_name(page);
+    adw_status_page_set_title(priv->status_page, _("All done!"));
+}
+
 static void
 gis_summary_page_constructed (GObject *object)
 {
@@ -268,6 +456,7 @@ gis_summary_page_constructed (GObject *object)
 
   update_distro_name (page);
   g_signal_connect (priv->start_button, "clicked", G_CALLBACK (done_cb), page);
+  g_signal_connect (priv->cancel_button, "clicked", G_CALLBACK (status_stack_reset_cb), page);
 
   gis_page_set_complete (GIS_PAGE (page), TRUE);
 
@@ -291,6 +480,9 @@ gis_summary_page_class_init (GisSummaryPageClass *klass)
 
   gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisSummaryPage, start_button);
   gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisSummaryPage, status_page);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisSummaryPage, spinner);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisSummaryPage, cancel_button);
+  gtk_widget_class_bind_template_child_private (GTK_WIDGET_CLASS (klass), GisSummaryPage, status_stack);
 
   page_class->page_id = PAGE_ID;
   page_class->locale_changed = gis_summary_page_locale_changed;
diff --git a/gnome-initial-setup/pages/summary/gis-summary-page.ui b/gnome-initial-setup/pages/summary/gis-summary-page.ui
index a0afa3a..efb306a 100644
--- a/gnome-initial-setup/pages/summary/gis-summary-page.ui
+++ b/gnome-initial-setup/pages/summary/gis-summary-page.ui
@@ -10,13 +10,54 @@
         </style>
 
         <child>
-          <object class="GtkButton" id="start_button">
-            <property name="use_underline">True</property>
+          <object class="GtkStack" id="status_stack">
             <property name="halign">center</property>
-            <style>
-              <class name="suggested-action"/>
-              <class name="pill"/>
-            </style>
+
+            <child>
+              <object class="GtkBox" id="spinner_box">
+               <property name="orientation">horizontal</property>
+               <property name="name">spinner</property>
+                <property name="spacing">10</property>
+                <property name="halign">center</property>
+                
+                <child>
+                  <object class="GtkSpinner" id="spinner">
+                    <property name="halign">center</property>
+                    <style>
+                      <class name="suggested-action"/>
+                      <class name="pill"/>
+                    </style>
+                  </object>
+                </child>
+
+                <child>
+                  <object class="GtkLabel" id="spinner_label">
+                    <property name="label" translatable="yes">Downloading additional language packs...</property>
+                    <property name="halign">center</property>
+                  </object>
+                </child>
+                
+                <child>
+                  <object class="GtkButton" id="cancel_button">
+                    <property name="label" translatable="yes">Cancel download</property>
+                    <property name="halign">center</property>
+                  </object>
+                </child>
+
+              </object>
+            </child>
+
+            <child>
+            <object class="GtkButton" id="start_button">
+                <property name="name">start</property>
+                <property name="use_underline">True</property>
+                <property name="halign">center</property>
+                <style>
+                  <class name="suggested-action"/>
+                  <class name="pill"/>
+                </style>
+              </object>
+            </child>
           </object>
         </child>
 
